app/src/main/java/com/example/codingtest1/MainActivity.kt

```diff
@@ -86,4 +94,29 @@ class MainActivity : AppCompatActivity() {
             .show()
     }
 
+    /**
+     * Save the list of contacts to SharedPreferences
+     */
+    private fun saveContactsToSharedPreferences() {
+        sharedPreferences = getSharedPreferences("shared preferences", MODE_PRIVATE)
+        val editor = sharedPreferences.edit()
+        val gson = Gson()
+        val json: String = gson.toJson(viewModel.contacts.value)
+        editor.putString("contacts", json)
+        editor.apply()
+    }
+
+    /**
+     * Load saved data from SharedPreferences and updates the ViewModel
+     */
+    private fun loadSavedDataFromSharedPreferences() {
+        // Load saved data from SharedPreferences and update ViewModel
+        sharedPreferences = getSharedPreferences("shared preferences", MODE_PRIVATE)
+        val gson = Gson()
+        val json: String? = sharedPreferences.getString("contacts", null)
+        val savedContacts: List<ContactsData> =
+            gson.fromJson(json, object : TypeToken<List<ContactsData>>() {}.type)
```

Comment:
Add null check for variable json to avoid encountering potential java.lang.NullPointerException
ex:
 if(json != null){